libs/api/domains/cms/src/lib/search/cmsSync.service.ts

```diff
@@ -187,4 +196,72 @@ export class CmsSyncService implements ContentSearchImporter<PostSyncOptions> {
     }
     return true
   }
+
+  /**
+   * Read all documents from an index and write their popularity scores to
+   * the documents with corresponding ids in the data array
+   *
+   * @param {string} index
+   * @param {Array<MappedData>} data
+   * @return {Array<MappedData>}
+   */
+  async migratePopularityScores(
+    index: string,
+    data: Array<MappedData>,
+  ): Promise<Array<MappedData>> {
+    if (!index) {
+      return data
+    }
+    // TODO: do a scrolling query or sumthin
+    const query = {
+      query: {
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        match_all: {},
+      },
+      size: 10000,
+      _source: ['popularityScore'],
+    }
+    const oldData = await this.elasticService.findByQuery<
+      SearchResponse<MappedData>,
+      unknown
+    >(index, query)
+
+    const oldScores = {}
+    oldData.body.hits?.hits.forEach((obj) => {
+      oldScores[obj._id] = obj._source.popularityScore
+    })
+
+    data.forEach((obj, idx) => {
+      if (obj._id in oldScores) {
+        data[idx].popularityScore = oldScores[obj._id]
+      }
+    })
+
+    return data
+  }
+
+  /**
+   * Returns the name of the most recent island-* index with the given locale,
+   * determined by creation date.
+   *
+   * @param {string} locale
+   * @return {string}
+   */
+  async getLatestIndex(locale: string): Promise<string> {
+    const client = await this.elasticService.getClient()
+
+    const indices = await client.cat.indices({
+      h: ['i', 'creation.date.string'],
+      s: 'creation.date:desc',
+    })
+
+    const results = indices.body.match(
+      new RegExp(`island-${locale}-[a-z0-9]+`, 'gi'),
+    )
+
+    if (results.length < 1) {
+      return null
+    }
+    return results[0]
+  }
```

Comment:
I'm not sure if this is the right approach.
Popularity score migration relates to index migration/creation not to index maintenance.
The cms sync is run periodically forcing us to have "when to run" logic and to dynamically find the target/source indexes.
We might get a cleaner approach by moving the migration logic into `island.is/apps/services/search-indexer/src/migrate/migrateElastic.ts` and not toughing the `popularityScore` inside imports.

Reply:
The popularity score migration is also done when doing `GET /sync` on the indexing service. Every time we want to update a document we need to do this score migration. It will read the old score of the documents from the index and put it in the documents in `importableData` before calling `client.bulk()` to write to ES. Otherwise the counter would be set to zero on every update, as the `index` function can not do partial updates to documents.

Reply:
Yes missed that, we could look into alternative methods of importing data into the index like update/upsert allowing us to keep existing popularity fields on documents.
If we make sure, all fields are always defined (should be this way already) we should be able to safely update existing documents even if they are setting a subset of fields to empty

Reply:
It would be cool to use something like upsert to persist the popularity score.

But from my understanding, the sync service does not know whether a document is being initialized or updated? (does the Contentful API provide this info?) We would need to pass that to the model mappers so that they know which fields to omit?

Maybe this is good enough for a first version? :)

Reply:
I'm also wondering how permanent this feature would be? If we are thinking about presenting traffic data like [gov.uk](https://www.gov.uk/performance/site-activity) does, as @eirikurn mentioned, then we could get the hit count from there.

Reply:
We need the popularityScore inside the document to be able to query by it so I see this as a permanent feature.
The challenges looks similar to me, I see the main difference in how the popularity score is updated in the document.

I believe adding `"doc_as_upsert":true` to the update query makes is behave as create or update (I haven't tested this).
If we can prevent the **sync destroys popularity score**&trade; issue we can handle this as a migration to be run only when new indexes are created

Reply:
This sounds good. I have moved this logic to the migration scripts, among other things (like using painless scripts and upserts instead of insert)